crates/walrus-indexer/src/storage.rs

```diff
@@ -0,0 +1,471 @@
+    fn to_key(&self) -> Vec<u8> {
+        // Simply use the raw ObjectID bytes (32 bytes).
+        self.object_id.as_ref().to_vec()
+    }
+
+    /// Parse from byte representation.
+    #[allow(unused)]
+    fn from_key(bytes: &[u8]) -> Result<Self> {
+        if bytes.len() != 32 {
+            return Err(anyhow::anyhow!(
+                "Invalid object_id length: expected 32 bytes"
+            ));
+        }
+        Ok(Self {
+            object_id: ObjectID::from_bytes(bytes)?,
+        })
+    }
+}
+
+impl std::fmt::Display for ObjectIndexKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.object_id)
+    }
+}
+
+/// Key structure for quilt patch index.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct QuiltPatchIndexKey {
+    /// The blob ID of the quilt patch.
+    patch_blob_id: BlobId,
+    /// The object ID of the containing quilt.
+    object_id: ObjectID,
+}
+
+impl std::fmt::Display for QuiltPatchIndexKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}/{}", self.patch_blob_id, self.object_id)
+    }
+}
+
+impl QuiltPatchIndexKey {
+    fn new(patch_blob_id: BlobId, object_id: ObjectID) -> Self {
+        Self {
+            patch_blob_id,
+            object_id,
+        }
+    }
+
+    /// Convert to optimized byte representation for RocksDB key.
+    /// Format: [patch_blob_id_bytes(32 bytes)] \x00 [object_id_bytes(32 bytes)].
+    /// \x00 is added here for future validation marker, e.g., if a quilt patch blob id is invalid,
+    /// we set the \x00 to \x01 to avoid reading it.
+    fn to_key(&self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+
+        // Add BlobId bytes (32 bytes).
+        bytes.extend_from_slice(self.patch_blob_id.as_ref());
+
+        // Add \x00 separator.
+        bytes.push(0x00);
+
+        // Add ObjectID bytes (32 bytes).
+        bytes.extend_from_slice(self.object_id.as_ref());
+
+        bytes
+    }
+
+    /// Parse from byte representation.
+    fn from_key(bytes: &[u8]) -> Result<Self> {
+        if bytes.len() != 65 {
+            return Err(anyhow::anyhow!(
+                "invalid key length: expected 65 bytes (32 + 1 + 32)"
+            ));
+        }
+
+        // Check for separator at position 32.
+        if bytes[32] != 0x00 {
+            return Err(anyhow::anyhow!("invalid quilt patch blob id"));
+        }
+
+        // Extract BlobId (first 32 bytes).
+        let blob_id_bytes: [u8; 32] = bytes[0..32].try_into()?;
+        let patch_blob_id = BlobId(blob_id_bytes);
+
+        // Extract ObjectID (last 32 bytes).
+        let object_id = ObjectID::from_bytes(&bytes[33..65])?;
+
+        Ok(Self {
+            patch_blob_id,
+            object_id,
+        })
+    }
+}
+
+// For scanning all entries with a specific blob_manager and identifier.
+fn get_primary_index_bounds(blob_manager: &ObjectID, identifier: &str) -> (Vec<u8>, Vec<u8>) {
```

Comment:
Nit (can be created in a followup PR): these helper functions can use some tests.